### ISSUE TITLE
Cannot survive from isValid(...) hang because IS_VALID_TIMEOUT in DefaultConnectionTester is hard coded to 0 

### ISSUE THREAD

**Issue by james-hu** [swaldman/c3p0#48]
Sometimes, maybe due to issues in Microsoft JDBC driver and/or Microsoft Azure service, isValid(..) method just hangs (or cannot return within several minutes). In that case, dropping the connection and creating another one seems to be better than waiting infinitely. 

I noticed that it is hard coded in DefaultConnectionTester.java that 
final static int    IS_VALID_TIMEOUT       = 0;

So the suggestion is the make this value configuration so that if needed, we can override it with values like 30 seconds or something.

Below is the dump of the hanging thread:

   java.lang.Thread.State: RUNNABLE
	at java.net.SocketInputStream.socketRead0(Native Method)
	at java.net.SocketInputStream.read(SocketInputStream.java:150)
	at java.net.SocketInputStream.read(SocketInputStream.java:121)
	at com.microsoft.sqlserver.jdbc.TDSChannel$ProxyInputStream.readInternal(IOBuffer.java:895)
	at com.microsoft.sqlserver.jdbc.TDSChannel$ProxyInputStream.read(IOBuffer.java:883)
	at sun.security.ssl.InputRecord.readFully(InputRecord.java:465)
	at sun.security.ssl.InputRecord.read(InputRecord.java:503)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:954)
	- locked <0x000000054692fcf8> (a java.lang.Object)
	at sun.security.ssl.SSLSocketImpl.readDataRecord(SSLSocketImpl.java:911)
	at sun.security.ssl.AppInputStream.read(AppInputStream.java:105)
	- locked <0x0000000546941688> (a sun.security.ssl.AppInputStream)
	at com.microsoft.sqlserver.jdbc.TDSChannel.read(IOBuffer.java:1782)
	at com.microsoft.sqlserver.jdbc.TDSReader.readPacket(IOBuffer.java:4838)
	- locked <0x00000006ac836c98> (a com.microsoft.sqlserver.jdbc.TDSReader)
	at com.microsoft.sqlserver.jdbc.TDSCommand.startResponse(IOBuffer.java:6150)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.doExecuteStatement(SQLServerStatement.java:797)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement$StmtExecCmd.doExecute(SQLServerStatement.java:696)
	at com.microsoft.sqlserver.jdbc.TDSCommand.execute(IOBuffer.java:5696)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.executeCommand(SQLServerConnection.java:1716)
	- locked <0x0000000546941bf0> (a java.lang.Object)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeCommand(SQLServerStatement.java:180)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeStatement(SQLServerStatement.java:155)
	at com.microsoft.sqlserver.jdbc.SQLServerStatement.executeQueryInternal(SQLServerStatement.java:631)
	at com.microsoft.sqlserver.jdbc.SQLServerConnection.isValid(SQLServerConnection.java:3369)
	at com.mchange.v2.c3p0.impl.DefaultConnectionTester$2.activeCheckConnectionNoQuery(DefaultConnectionTester.java:128)
	at com.mchange.v2.c3p0.impl.DefaultConnectionTester$3.activeCheckConnectionNoQuery(DefaultConnectionTester.java:172)
	at com.mchange.v2.c3p0.impl.DefaultConnectionTester.activeCheckConnection(DefaultConnectionTester.java:251)
	at com.mchange.v2.c3p0.AbstractConnectionTester.activeCheckConnection(AbstractConnectionTester.java:79)
	at com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool$1PooledConnectionResourcePoolManager.testPooledConnection(C3P0PooledConnectionPool.java:503)

**Comment by swaldman** [comment]
(hi. i'm sorry. this slipped by me. it sounds like a reasonable suggestion to make it configurable. i'll look into that!)

**Comment by swaldman** [comment]
hi,

there's now a 0.9.5.1-SNAPSHOT release which implements a config param `com.mchange.v2.c3p0.impl.DefaultConnectionTester.isValidTimeout`.

if you're still interested in this, please give this release a shot. i'd appreciate any feedback before i call 0.9.5.1 a release.

many thanks, and sorry for the slow work!

**Comment by james-hu** [comment]
Hi swaldman,
Thanks for looking into this and making improvement in such a short time. I will have a try and let you know the result.

**Comment by james-hu** [comment]
Hi swaldman,

There are good news and bad news, and more:

1) The newly added configuration parameter works. isValid(...) method of the JDBC driver did time out during the test.

2) It seems that I have to put the configuration in a global properties file, like c3p0.properties. I don't know how to set the configuration in the use case of DataSources.pooledDataSource(...). I do wish it can be supported in this case.

3) In my specific case, the actual time out is 15 minutes, although I set 19 seconds in the configuration file and I can see "19 seconds" in the log message. I am connecting from a virtual machine inside Microsoft Azure to an Azure SQL database. Probably this is an issue within Microsoft's JDBC driver or even below that.

4) In my specific case, after first attempt to acquire a valid database connection failed due to time out, c3p0 tried to acquire another connection, but probably because of issues within JDBC driver or even below that, the second attempt resulted in another time out after 15 minutes, and c3p0 tried the third attempt... So the application is still waiting and there is no usable connection can be returned. I am not sure whether this can be solved in c3p0 or not.

So, the bottom line is, the configuration parameter works. And I do appreciate your work on that. This might not solve my problems directly but surely may be useful for others.

James

**Comment by swaldman** [comment]
I can't explain while 19 seconds became 15 minutes... that doesn't sound good. I'm pretty sure, if you set 19 and you see 19 in the logs, connection.isValid( 19 ) is what is getting called. So, I'll guess it's somewhere in the driver.

Unfortunately, there'd be no way to let DefaultConnectionTester accept a timeout on a case-by-case basis without a disruptive update to the ConnectionTester API, since multiple DataSources may share a ConnectionTester. Any customizations of DefaultConnectionTester have to be configured globally, thus the properties style configuration parameter.

If you do want to customize timeouts on a case-by-case basis, the simplest way would be to write custom ConnectionTesters. It turns out, that's not as simple as i would like it to be, so I've written a base class for you, `com.mchange.v2.c3p0.util.IsValidOnlyConnectionTester`. Now it is trivial to define new classes configured with whatever timeout you'd like.

```java
package mypkg;

import com.mchange.v2.c3p0.util.IsValidOnlyConnectionTester;

public final class IsValidOnlyConnectionTester30 extends IsValidOnlyConnectionTester
{
    protected int getIsValidTimeout() { return 30; }
}
```

Make sure your classes are compiled and bundled with your application. Now, in your call to `DataSources.pooledDataSource(...)`, you can include in your `overrideProps` you can put `("connectionTesterClassName", "mypkg. IsValidOnlyConnectionTester30")`, and a timeout of 30 secs will be configured for that DataSource. You can make as many timing-out subclasses as you'd like (each with a different name), and set different `connectionTesterClassName` values for different DataSource if you wish.

The example above is [embedded](https://github.com/swaldman/c3p0/blob/master/src/java/com/mchange/v2/c3p0/example/IsValidOnlyConnectionTester30.java) in the `example` package of the library now. The new abstract base class is [here](https://github.com/swaldman/c3p0/blob/master/src/java/com/mchange/v2/c3p0/util/IsValidOnlyConnectionTester.java).

I hope this is a little helpful. Good luck!

**Comment by james-hu** [comment]
This is a very practical solution. Thanks!

**Comment by swaldman** [comment]
OK. So this stuff is all included in 0.9.5.1, which is now a release. Thanks for your help and feedback!
